Ingesters: Delete blocks relative to when they were uploaded to storage

jesusvazquez opened the pull request:
This PR changes the logic of when blocks will be deleted on
ingesters. Before it was relative to the newest block for the tenant
minus a certain retention. Now it depends on whether or not shipping is
enabled.
- If it is enabled, deletion will be relative to when the block was
  uploaded minus a certain retention.
- If it is disabled, deletion will be relative to when the block was
  created minos a certain retention.

We do this for two reasons:
- It will allow us to lower the retention on ingesters making them
  ligther.
- If the out-of-order window is bigger than the retention time, with the
  previous logic blocks would be deleted right after upload, leaving a
time where data would not be queryable.

pracucci commented:
Have you considered making it based on when the block is created? You can read the creation timestamp from the block ULID.

jesusvazquez commented:
@pracucci ULID can give me that information but I actually need to know it was shipped because I have to wait until its shipped and re-scanned, from ingesters I know the first part with this implementation and I can have an estimation as a constant for the second part.

If I did it relative to block creation time maybe its shipped late for whatever reason i.e errors uploading to gcs, and whenever it got uploaded it could also get deleted. There could be a time there were data disappears for queries until blocks are re-scanned by the store gateways.

colega commented:
I think that doing it based on shipping time is safer: in most case it will be the same, but as @jesusvazquez mentions, in case of a block storage outage for whatever reason, we'll keep the block for enough time.

colega commented on pkg/ingester/shipper_test.go:
```diff
@@ -305,27 +305,27 @@ func TestReadMetaFile(t *testing.T) {
 
-		_, err := readShipperMetaFile(dpath)
-		fpath := filepath.Join(dpath, shipperMetaFilename)
+		_, err := readThanosShipperMetaFile(dpath)
```
We should rename this test to `TestReadThanosMetaFile` and replicate the same test for the new Mimir's meta.

colega commented on pkg/ingester/user_tsdb.go:
```diff
@@ -245,9 +246,11 @@ func (u *userTSDB) blocksToDelete(blocks []*tsdb.Block) map[ulid.ULID]struct{} {
-		if _, ok := deletable[shippedID]; ok {
-			result[shippedID] = struct{}{}
+	for blockID, blockShippedTime := range shippedBlocks {
+		if blockShippedTime.Before(time.Now().Add(-u.blockMinRetention)) {
```
I'd extract `time.Now().Add(-u.blockMinRetention)` to a variable called `deadline` or similar.

56quarters commented:
Since this changes the mental model of how blocks are managed on ingesters, please make sure you update any relevant docs and runbooks.

jesusvazquez commented:
Hey @56quarters, I updated the docs and the reference help as well as added an entry in the CHANGELOG and updated the description of the `blocks-storage.tsdb.retention-period` flag. 

Do you know of any other doc pointers I should update?

In the meantime I'll set this PR as ready to review.

colega commented:
This looks good to me, also the change makes a lot of sense. I did some pair programming on this though, so we'd need an approval from someone who's not me.

pstibrany commented:
Shipping is an optional feature, and there are users who run Mimir with shipping disabled. Is that case properly handled by the PR? (Just started the review)

pstibrany commented on pkg/ingester/shipper.go:
```diff
@@ -152,34 +145,45 @@ func (s *Shipper) Sync(ctx context.Context) (uploaded int, err error) {
 		if ok {
-			meta.Uploaded = append(meta.Uploaded, m.ULID)
-			uploaded++ // the last upload must have failed, report the block as if it was uploaded successfully now
+			// TODO Whishlist would be great to have a file in the bucket telling us when the block was uploaded.
```
That would be last modified time of `meta.json` file (because it's uploaded as last), which can be obtained by calling `bucket.Attributes` instead of `bucket.Exists`. Also we try not to leave "wishlist" TODOs around.

jesusvazquez replied:
Decided to be conservative and simply say it was rencetly uploaded. Removed the wishlist and TIL about bucket.Attributes.

pstibrany commented on pkg/ingester/shipper.go:
```diff
@@ -233,41 +237,68 @@ func (s *Shipper) blockMetasFromOldest() (metas []*metadata.Meta, _ error) {
+// shipperMeta defines the format thanos.shipper.json file that the shipper places in the data directory.
+type shipperMeta struct {
+	Version int            `json:"version"`
+	Shipped []shippedBlock `json:"shipped"`
```
What about `Shipped map[ulid.ULID]time.Time` instead? (or perhaps `model.Time`)

pstibrany replied:
Using `Shipped map[ulid.ULID]model.Time` looks like this:

```
{
  "version": 1,
  "shipped": {
    "01GPGRWYPY0000000000000000": "2023-01-11T16:59:07.486415+01:00"
  }
}
```

and with `model.Time`:
```
{
  "version": 1,
  "shipped": {
    "01GPGRZ5H70000000000000000": 1673452820.007
  }
}
```

I'd prefer the last version.

jesusvazquez replied:
Okey, used last version.

pstibrany commented on pkg/ingester/shipper.go:
```diff
@@ -286,18 +317,71 @@ func writeShipperMetaFile(logger log.Logger, dir string, meta *shipperMeta) erro
-// readShipperMetaFile reads the given meta from <dir>/thanos.shipper.json.
-func readShipperMetaFile(dir string) (*shipperMeta, error) {
+// writeThanosShipperMetaFile writes the given meta into <dir>/thanos.shipper.json.
+func writeThanosShipperMetaFile(logger log.Logger, dir string, meta *thanosShipperMeta) error {
```
`writeShipperMetaFile` and `writeThanosShipperMetaFile` have almost the same body, can we remove this duplication?

jesusvazquez replied:
`writeThanosShipperMetaFile` will get removed in a future release plus it also gets called from `writeShipperMetaFile`, I could substract the file creation meta encoding and os file creation to its own method but then it would only have one caller in the future

pstibrany commented on pkg/ingester/shipper.go:
```diff
@@ -286,18 +317,71 @@ func writeShipperMetaFile(logger log.Logger, dir string, meta *shipperMeta) erro
 		return err
 	}
+
+	// Keep backwards compatibility with the previous Mimir version.
```
This is fine, but why not finish writing mimir-file first (by doing the rename)?

pstibrany commented on pkg/ingester/shipper.go:
```diff
@@ -233,41 +237,68 @@ func (s *Shipper) blockMetasFromOldest() (metas []*metadata.Meta, _ error) {
+	shippedBlocks := make(map[ulid.ULID]time.Time, len(meta.Uploaded))
 	for _, blockID := range meta.Uploaded {
-		shippedBlocks[blockID] = struct{}{}
+		shippedBlocks[blockID] = ulid.Time(blockID.Time()).Add(estimatedShippingDuration)
```
Shall we set it to `time.Now()` instead? When shipper runs next time, it will use store `now` to `mimir.shipper.json` file anyway (or timestamp of meta.json file), so next call to `readShippedBlocks` will use that value, and we don't need to provide any "estimated" ship time.

pstibrany reviewed (commented):
Looks good, but I don't see above-mentioned case when shipping is disabled handled in the PR. With shipping disabled, perhaps we can check block creation time against the "deadline".

I'd also suggest adding test for "shipping disabled" case.

pstibrany commented on pkg/ingester/shipper.go:
```diff
@@ -233,41 +230,63 @@ func (s *Shipper) blockMetasFromOldest() (metas []*metadata.Meta, _ error) {
 }
 
-func readShippedBlocks(dir string) (map[ulid.ULID]struct{}, error) {
+func readShippedBlocks(dir string) (map[ulid.ULID]time.Time, error) {
```
[nit] We can simplify shipper code a bit if we return `map[ulid.ULID]model.Time` here.

pstibrany commented on pkg/ingester/shipper.go:
```diff
@@ -152,34 +144,39 @@ func (s *Shipper) Sync(ctx context.Context) (uploaded int, err error) {
-		meta.Uploaded = append(meta.Uploaded, m.ULID)
-		uploaded++
+
+		meta.Shipped[m.ULID] = model.TimeFromUnixNano(time.Now().UnixNano())
```
```suggestion
		meta.Shipped[m.ULID] = model.Now()
```

pstibrany commented on pkg/ingester/shipper.go:
```diff
@@ -152,34 +144,39 @@ func (s *Shipper) Sync(ctx context.Context) (uploaded int, err error) {
+			// We decide to be conservative here and assume it was just recently uploaded.
+			// It would be very rare to have blocks uploaded but not tracked in the shipper meta file.
+			// This could happen if process crashed while uploading the block.
+			meta.Shipped[m.ULID] = model.TimeFromUnixNano(time.Now().UnixNano())
```
[nit] We can use `model.Now()`

pstibrany commented on pkg/ingester/user_tsdb.go:
```diff
@@ -85,8 +86,9 @@ type userTSDB struct {
-	shippedBlocks    map[ulid.ULID]struct{}
+	shippedBlocksMtx  sync.Mutex
+	shippedBlocks     map[ulid.ULID]time.Time
+	blockMinRetention time.Duration
```
[nit] I'd separate this field from previous two, as now it looks like it's protected by `shippedBlocksMtx`, which is not true. (We can eg. move it before `shippedBlocks*` fields).

pstibrany commented on pkg/ingester/user_tsdb.go:
```diff
@@ -237,19 +239,35 @@ func (u *userTSDB) blocksToDelete(blocks []*tsdb.Block) map[ulid.ULID]struct{} {
+
+		for blockID, blockShippedTime := range shippedBlocks {
+			if blockShippedTime.Before(deadline) {
+				if _, ok := deletable[blockID]; ok {
```
[nit] Now that `shippedBlocks` is a map, we could iterate over `deletable` and check if it was shipped. The logic would be clearer that way, but it's not a blocker.

pstibrany commented on pkg/ingester/user_tsdb.go:
```diff
@@ -237,19 +239,35 @@ func (u *userTSDB) blocksToDelete(blocks []*tsdb.Block) map[ulid.ULID]struct{} {
-			result[shippedID] = struct{}{}
+	for _, b := range blocks {
+		blockID := b.Meta().ULID
+		blockCreationTime := util.TimeFromMillis(int64(blockID.Time()))
```
[nit] We can use `time.UnixMilli` instead of `util.TimeFromMillis`.

pstibrany commented on pkg/ingester/user_tsdb.go:
```diff
@@ -237,19 +239,35 @@ func (u *userTSDB) blocksToDelete(blocks []*tsdb.Block) map[ulid.ULID]struct{} {
-	for shippedID := range shippedBlocks {
-		if _, ok := deletable[shippedID]; ok {
-			result[shippedID] = struct{}{}
+	for _, b := range blocks {
```
Why not iterate over `deletable` only?

pstibrany reviewed (approved):
Thank you, PR looks good to me. I've left some non-blocking comments.

pstibrany reviewed (approved):
Thank you very much!

pracucci commented on pkg/ingester/shipper.go:
```diff
@@ -100,38 +102,28 @@ func NewShipper(
 		// If we encounter any error, proceed with an empty meta file and overwrite it later.
 		// The meta file is only used to avoid unnecessary bucket.Exists call,
 		// which are properly handled by the system if their occur anyway.
-		if !os.IsNotExist(err) {
```
Why this was removed?

pracucci replied:
I think because can't happen because this specific error is swallowed by `readThanosShippedBlocks()`.

jesusvazquez commented:
Thank you for the reviews @pstibrany !! :tada:

pracucci commented on pkg/ingester/shipper.go:
```diff
@@ -233,41 +230,63 @@ func (s *Shipper) blockMetasFromOldest() (metas []*metadata.Meta, _ error) {
 	Uploaded []ulid.ULID `json:"uploaded"`
 }
 
+// shipperMeta defines the format thanos.shipper.json file that the shipper places in the data directory.
```
```suggestion
// shipperMeta defines the format mimir.shipper.json file that the shipper places in the data directory.
```

jesusvazquez replied:
Addressed :+1:

pracucci commented on pkg/ingester/shipper.go:
```diff
@@ -279,6 +298,45 @@ func writeShipperMetaFile(logger log.Logger, dir string, meta *shipperMeta) erro
+	}
+
+	// Keep backwards compatibility with the previous Mimir version.
+	// TODO Remove in Mimir 2.7.0
```
Please 2.7 with 2.8 across the file. We'll announce in 2.6, so we can remove starting 2.8.

pracucci commented on pkg/ingester/shipper.go:
```diff
@@ -100,38 +102,28 @@ func NewShipper(
+		shippedBlocks = map[ulid.ULID]time.Time{}
 	}
 
-	// Reset the uploaded slice, so we can rebuild it only with blocks that still exist locally.
```
[nit] I think this comment was helpful to clarify why we reset Shipped to an empty map.

jesusvazquez replied:
Added comment right above where we initialize the empty slice in case of error

jesusvazquez commented on pkg/ingester/shipper.go:
```diff
@@ -233,41 +233,63 @@ func (s *Shipper) blockMetasFromOldest() (metas []*metadata.Meta, _ error) {
+}
+
+// readThanosShippedBlocks reads the legacy thanos.shipper.json file.
+// TODO Remove in Mimir 2.7.0
```
Addressed as part of https://github.com/grafana/mimir/pull/3816/commits/03a700f47d20a45bc8bc86789088b9e4c61b0ce2
